### ISSUE TITLE
Two-node DDP shows reproducible FP32 loss divergence between CPU/Gloo and CUDA/NCCL with large lr and resnet18

### ISSUE THREAD

**Issue by whitephilomel** [pytorch/pytorch#187279]
### 🐛 Describe the bug

A two-node `DistributedDataParallel` training job produces substantially different results when the same deterministic FP32 workload is executed with:
- CPU + Gloo
- CUDA + NCCL

Both arms use the same ResNet-18 architecture, initial seed, per-rank synthetic input tensors, labels, optimizer settings, world size, and number of optimize steps. The model is initialized from the same seed before being moved to the selected device.

After 20 AdamW steps, rank 1 reports:

- CPU/Gloo final loss: `2.5815484523773193`
- CUDA/NCCL final loss: `4.375443458557129`
- absolute difference: `1.7938950061798096`
- relative difference: `40.999%`

The final gradient norm also changes from `0.8940557241439819` on CPU to `2.136383533477783` on CUDA, a relative difference of approximately `58.15%`.

This is not an AMP, TF32, data-order, dataset-download, or cuBLAS nondeterminism comparison:

- precision is FP32 in both arms;
- AMP and GradScaler are disabled;
- TF32 is disabled;
- `torch.use_deterministic_algorithms(True, warn_only=True)` is enabled;
- cuDNN benchmark mode is disabled;
- `CUBLAS_WORKSPACE_CONFIG=:4096:8` is set before Python starts;
- inputs and labels are generated on CPU from fixed per-rank seeds and reused
  unchanged between the two arms;
- no dataset or DataLoader is involved.

Both distributed runs complete all 20 steps and exit normally. The problem is the reproducible numerical divergence, not a process crash.
## Reproduction

The following files are attached:

```text
minimal_repro_ddp_differential.py
pytorch_official_repro_config.json
```
You can get the repro code from https://gist.github.com/whitephilomel/92ca26f8b082b62c33ee5a5becb63ca2

The reproducer requires two machines with one process and one GPU per machine. Run it from the directory containing both files.

The relevant configuration is:

```text
model                     = resnet18
optimizer                 = AdamW
learning rate             = 1.0
weight decay              = 0.0
precision                 = fp32
AMP                       = disabled
TF32                      = disabled
gradient clipping         = 5.0
per-rank batch size       = 8
world size                = 2
steps                     = 20
seed                      = 0
DDP broadcast_buffers     = false
DDP bucket_cap_mb         = 128
DDP static_graph          = false
DDP gradient bucket view  = false
```

The complete configuration is contained in `pytorch_official_repro_config.json`.

## Actual behavior

The most recent confirmed run of `minimal_repro_ddp_differential.py`, performed with `CUBLAS_WORKSPACE_CONFIG=:4096:8`, produced:

| rank | CPU/Gloo loss | CUDA/NCCL loss | absolute difference | relative difference | CPU grad norm | CUDA grad norm |
|---:|---:|---:|---:|---:|---:|---:|
| 0 | 2.9489951133728027 | 4.01106071472168 | 1.062065601348877 | 26.478% | 0.8940557241439819 | 2.136383533477783 |
| 1 | 2.5815484523773193 | 4.375443458557129 | 1.7938950061798096 | 40.999% | 0.8940557241439819 | 2.136383533477783 |

Additional final-state differences:

| metric | CPU/Gloo | CUDA/NCCL | relative difference |
|---|---:|---:|---:|
| gradient norm | 0.8940557241439819 | 2.136383533477783 | 58.151% |
| logits sum, rank 1 | 327.8002624511719 | 283.2908935546875 | 13.578% |
| parameter sum | -311665.84375 | -263532.53125 | 15.444% |

The decisive summary emitted on rank 1 is:

```json
{
  "status": "DIVERGENCE_REPRODUCED",
  "mode": "cpu_vs_gpu",
  "rank": 1,
  "reason": "arms diverged: final_loss",
  "final_loss": {
    "cpu": 2.5815484523773193,
    "gpu": 4.375443458557129,
    "abs_diff": 1.7938950061798096,
    "rel_diff": 0.409991586720532
  },
  "final_grad_norm": {
    "cpu": 0.8940557241439819,
    "gpu": 2.136383533477783,
    "abs_diff": 1.2423278093338013,
    "rel_diff": 0.5815097288788014
  }
}
```

## Expected behavior

Small floating-point differences between CPU and CUDA are expected, and bitwise equality is not required. However, with fixed inputs, fixed initialization, deterministic execution controls, FP32 computation, and only 20 optimizer steps, the two DDP arms should remain numerically close enough that the final loss and gradient norm do not diverge by approximately 41% and 58%, respectively.

If this degree of backend-dependent drift is expected for this supported configuration, it would be useful to document the expected numerical bounds and the operations responsible for the amplification.

### Versions

PyTorch: 2.9.1+cu128
Python: 3.11.15
CUDA reported by PyTorch: 12.8
NCCL: 2.27.5+cuda12.9
OS: Linux 5.15.0-113-generic x86_64, glibc 2.35
GPU: Tesla V100S-PCIE-32GB, one GPU per machine
Distributed topology: two machines, world_size=2, one process per machine
CPU：Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy

**Comment by cx2009** [comment]
Is this blocking training or just intermittent behavior?

**Comment by whitephilomel** [comment]
Hi @cx2009 ,this issue is reproducible, not an intermittent behavior.

**Comment by arkadip-maitra** [comment]
This seems like a CPU vs CUDA numerical drift and not a ddp/nccl bug. In single device setup also a numerical drift is there
